internal/storage/field_stats.go

```diff
@@ -0,0 +1,359 @@
+	case schemapb.DataType_Int64:
+		data := msgs.(*Int64FieldData).Data
+
+		// return error: msgs must has one element at least
+		if len(data) < 1 {
+			return
+		}
+
+		b := make([]byte, 8)
+		for _, int64Value := range data {
+			pk := NewInt64FieldValue(int64Value)
+			stats.UpdateMinMax(pk)
+			common.Endian.PutUint64(b, uint64(int64Value))
+			stats.BF.Add(b)
+		}
+	case schemapb.DataType_VarChar:
+		data := msgs.(*StringFieldData).Data
+
+		// return error: msgs must has one element at least
+		if len(data) < 1 {
+			return
+		}
+
+		for _, str := range data {
+			pk := NewVarCharFieldValue(str)
+			stats.UpdateMinMax(pk)
+			stats.BF.AddString(str)
+		}
+	default:
+		// TODO::
+	}
+}
+
+func (stats *FieldStats) Update(pk ScalarFieldValue) {
+	stats.UpdateMinMax(pk)
+	switch stats.Type {
+	case schemapb.DataType_Int64:
+		data := pk.GetValue().(int64)
+		b := make([]byte, 8)
+		common.Endian.PutUint64(b, uint64(data))
+		stats.BF.Add(b)
+	case schemapb.DataType_VarChar:
+		data := pk.GetValue().(string)
+		stats.BF.AddString(data)
+	default:
+		// todo support vector field
+	}
+}
+
+// UpdateMinMax update min and max value
+func (stats *FieldStats) UpdateMinMax(pk ScalarFieldValue) {
+	if stats.Min == nil {
+		stats.Min = pk
+	} else if stats.Min.GT(pk) {
+		stats.Min = pk
+	}
+
+	if stats.Max == nil {
+		stats.Max = pk
+	} else if stats.Max.LT(pk) {
+		stats.Max = pk
+	}
+}
+
+func NewFieldStats(fieldID int64, pkType schemapb.DataType, rowNum int64) (*FieldStats, error) {
+	return &FieldStats{
+		FieldID: fieldID,
+		Type:    pkType,
+		BF:      bloom.NewWithEstimates(uint(rowNum), paramtable.Get().CommonCfg.MaxBloomFalsePositive.GetAsFloat()),
+	}, nil
+}
+
+// FieldStatsWriter writes stats to buffer
+type FieldStatsWriter struct {
+	buffer []byte
+}
+
+// GetBuffer returns buffer
+func (sw *FieldStatsWriter) GetBuffer() []byte {
+	return sw.buffer
+}
+
+// GenerateList writes Stats slice to buffer
+func (sw *FieldStatsWriter) GenerateList(stats []*FieldStats) error {
+	b, err := json.Marshal(stats)
+	if err != nil {
+		return err
+	}
+	sw.buffer = b
+	return nil
+}
+
+// GenerateByData writes Int64Stats or StringStats from @msgs with @fieldID to @buffer
+func (sw *FieldStatsWriter) GenerateByData(fieldID int64, pkType schemapb.DataType, msgs ...FieldData) error {
+	statsList := make([]*FieldStats, 0)
+	for _, msg := range msgs {
+		stats := &FieldStats{
+			FieldID: fieldID,
+			Type:    pkType,
+			BF:      bloom.NewWithEstimates(uint(msg.RowNum()), paramtable.Get().CommonCfg.MaxBloomFalsePositive.GetAsFloat()),
+		}
+
+		stats.UpdateByMsgs(msg)
+		statsList = append(statsList, stats)
+	}
+	return sw.GenerateList(statsList)
+}
+
+// FieldStatsReader reads stats
+type FieldStatsReader struct {
+	buffer []byte
+}
+
+// SetBuffer sets buffer
+func (sr *FieldStatsReader) SetBuffer(buffer []byte) {
+	sr.buffer = buffer
+}
+
+// GetFieldStats returns buffer as FieldStats
+func (sr *FieldStatsReader) GetFieldStats() (*FieldStats, error) {
```

Comment:
I'm confused.
When is the time I should call GetFieldStats and GetFieldStatsList?

I fully believe we should not bring in new concepts like FieldStatsList and FieldStats.

FieldStatsReader and FieldStatsWriter should work on FieldStats slice. then only difference is the slice has one element and multiple elements.